plat/common/arm/mte.c

```diff
@@ -0,0 +1,117 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2022, Michalis Pappas <mpappas@fastmail.fm>.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <arm/arm64/cpu.h>
+#include <stddef.h>
+#include <uk/plat/mte.h>
+#include <uk/arch/lcpu.h>
+#include <uk/arch/random.h>
+#include <uk/arch/traps.h>
+#include <uk/event.h>
+
+void *mte_tag_contig(void *ptr, size_t size)
+{
+	uint64_t addr = (uint64_t)ptr;
+
+	UK_ASSERT(!(size % MTE_TAG_GRANULE));
+
+	addr = mte_insert_random_tag(addr);
+
+	for (unsigned int i = 0; i < size; i += MTE_TAG_GRANULE)
+		mte_store_alloc(addr + i);
+
+	return (void *)addr;
+}
+
+int mte_init(void)
+{
+	__u64 reg;
+	__u64 seed;
+	unsigned int mte_version;
+
+	mte_version = (SYSREG_READ(ID_AA64PFR1_EL1) >>
+			ID_AA64PFR1_EL1_MTE_SHIFT) &
+			ID_AA64PFR1_EL1_MTE_MASK;
+
+	if (mte_version < ARM64_FEAT_MTE2) {
+		uk_pr_err("FEAT_MTE2 is not implemented\n");
+		return -1;
```

Comment:
Hmm, is there an errno for missing hardware feature?

Reply:
In general I'm also in favor of errno, but given that MTE is a security feature perhaps we don't want to handle an initialization error gracefully? Of course we can return values from errno and let the platform choose how to handle it. I will change to `-ENOTSUP`.

Btw, maybe I'm missing something but I see some inconsistency here. On the one hand we don't want `stdint` types in global headers (in my understanding to avoid libc dependency) and on the other hand we use `errno` values sparingly. I know we discussed that before (IIRC in the scope of the SMP API) and the outcome was something along the lines of _"by using errno numbers we don't mandate libc, the platform is free to define these values independently of a standard library"_. Do I remember correctly? Please correct me if I'm wrong. But if we go down that road, what stops someone from arguing that we can then also assume the same for `stdint` given that types are more fundamental than error numbers? 😬 Where is the line? 🤔 I'll add this topic to in my notes for discussion on plat-rearch.

Reply:
> In general I'm also in favor of errno, but given that MTE is a security feature perhaps we don't want to handle an initialization error gracefully?

Good point. I mean we do crash in the platform anyways, so technically it does not make a difference. 

My argument was that while the caller may not know the details, he generally has a better picture of the context. So, providing the caller enough information so that he can make an informed decision is usually a good idea, I think. And if a security feature is critical to the operational context, the caller will know and terminate anyways.

> Btw, maybe I'm missing something but I see some inconsistency here. On the one hand we don't want stdint types in global headers (in my understanding to avoid libc dependency)...

Yes, you are absolutely right. From my perspective, I am strongly in favor of adding a `uk/errno.h` with Linux compatible values (at least a sensible subset). The header would define the values if not provided by a C library. We could then include `uk/errno.h` everywhere, also in low-level platform code.

Reply:
Agreed, thanks 👍🏼 I updated `mte_init()` to return `-ENOTSUP` in the last iteration already, so I think we can resolve this one now. I put the topic of `errno.h` in my list for plat-rearch discussion.